include/exadg/grid/mapping_finite_element.h

```diff
@@ -0,0 +1,373 @@
+    dof_handler.distribute_mg_dofs();
+    VectorType grid_coordinates_fine_level;
+    fill_grid_coordinates_vector(*this->mapping, grid_coordinates_fine_level, dof_handler);
+
+    MGTransferMatrixFree<dim, Number> transfer;
+    transfer.build(dof_handler);
+    transfer.interpolate_to_mg(dof_handler,
+                               grid_coordinates_all_levels,
+                               grid_coordinates_fine_level);
+
+    for(unsigned int level = 0; level < n_levels; level++)
+    {
+      IndexSet relevant_dofs;
+      DoFTools::extract_locally_relevant_level_dofs(dof_handler, level, relevant_dofs);
+
+      grid_coordinates_all_levels_ghosted[level].reinit(
+        dof_handler.locally_owned_mg_dofs(level),
+        relevant_dofs,
+        grid_coordinates_all_levels[level].get_mpi_communicator());
+
+      grid_coordinates_all_levels_ghosted[level].copy_locally_owned_data_from(
+        grid_coordinates_all_levels[level]);
+
+      grid_coordinates_all_levels_ghosted[level].update_ghost_values();
+    }
+
+    AssertThrow(fe.element_multiplicity(0) == dim,
+                ExcMessage("Expected finite element with dim components."));
+
+    FE_Nothing<dim> fe_nothing;
+    FEValues<dim>   fe_values(*mapping,
+                            fe_nothing,
+                            QGaussLobatto<dim>(fe.degree + 1),
+                            update_quadrature_points);
+
+    MappingQCache<dim>::initialize(
+      dof_handler.get_triangulation(),
+      [&](const typename Triangulation<dim>::cell_iterator & cell_tria) -> std::vector<Point<dim>> {
+        unsigned int const level = cell_tria->level();
+
+        typename DoFHandler<dim>::cell_iterator cell(&cell_tria->get_triangulation(),
+                                                     level,
+                                                     cell_tria->index(),
+                                                     &dof_handler);
+
+        unsigned int const scalar_dofs_per_cell = Utilities::pow(fe.degree + 1, dim);
+
+        std::vector<Point<dim>> grid_coordinates(scalar_dofs_per_cell);
+
+        if(cell->level_subdomain_id() != numbers::artificial_subdomain_id)
+        {
+          fe_values.reinit(typename Triangulation<dim>::cell_iterator(cell));
+          std::vector<types::global_dof_index> dof_indices(fe.dofs_per_cell);
+          cell->get_mg_dof_indices(dof_indices);
+
+          // extract displacement and add to original position
+          for(unsigned int i = 0; i < scalar_dofs_per_cell; ++i)
+          {
+            grid_coordinates[i] =
+              fe_values.quadrature_point(this->hierarchic_to_lexicographic_numbering[i]);
+          }
+
+          for(unsigned int i = 0; i < dof_indices.size(); ++i)
+          {
+            std::pair<unsigned int, unsigned int> const id = fe.system_to_component_index(i);
+
+            if(fe.dofs_per_vertex > 0) // FE_Q
+            {
+              grid_coordinates[id.second][id.first] +=
```

Comment:
@kronbichler Adding into the vector is probably not wrong, but an assignment would be clearer?